tests/terraform_pytest/test_terraform_install.py

```diff
@@ -0,0 +1,97 @@
+"""Run Terraform plan/apply against each set of Terraform test files."""
+
+import os
+from pathlib import Path
+
+import pytest
+import tftest
+
+AWS_DEFAULT_REGION = os.getenv("AWS_DEFAULT_REGION", default="us-east-1")
+MOCKSTACK_HOST = os.getenv("MOCKSTACK_HOST", default="localhost")
+MOCKSTACK_PORT = "4566"
+
+MOCKSTACK_TF_FILENAME = "mockstack.tf"
+AWS_TF_FILENAME = "aws.tf"
+
+
+@pytest.fixture(scope="function")
+def tf_test_object(is_mock, tf_dir):
+    """Return function that will create tf_test object using given subdir."""
+
+    def make_tf_test(tf_module):
+        """Return a TerraformTest object for given module."""
+        tf_test = tftest.TerraformTest(tf_module, basedir=str(tf_dir), env=None)
+
+        # Use the appropriate endpoints, either for a simulated AWS stack
+        # or the real deal.
+        provider_tf = MOCKSTACK_TF_FILENAME if is_mock else AWS_TF_FILENAME
+
+        current_dir = Path(__file__).resolve().parent
+        tf_test.setup(extra_files=[str(Path(current_dir / provider_tf))])
+        return tf_test
+
+    return make_tf_test
+
+
+@pytest.fixture(scope="function")
+def tf_vars(is_mock):
+    """Return values for variables used for the Terraform apply.
+
+    Set the Terraform variables for the hostname and port to differentiate
+    between using "localhost" or the docker network name.
+    """
+    return (
+        {
+            "mockstack_host": MOCKSTACK_HOST,
+            "mockstack_port": MOCKSTACK_PORT,
+        }
+        if is_mock
+        else {}
+    )
+
+
+@pytest.fixture(scope="function")
+def apply_plan(tf_vars):
+    """Return function that can be invoked with tf_test parameter."""
+
+    def invoke_plan_and_apply(tf_test):
+        """Execute Terraform plan and apply."""
+        # Debugging info:  tftest's plan() will raise an exception if the
+        # return code is 1.  Otherwise, it returns the text of the plan
+        # output.  Adding an argument of "output=True" will return an object
+        # that has members for outputs, resources, modules and variables.
+        # For this test, we're simply looking for pass or fail.
+        #
+        # tftest's apply() will also raise an exception if the return code
+        # is 1.  Otherwise, it returns the output as plain text.
+        try:
```

Comment:
This is the "try" block I am referring to.  If the apply() fails, the "except" below executes the "tf_test.destroy(tf_vars=tf_vars)".

Reply:
Ahh, ok. I missed this one line in that change. Sorry about that. However, this still doesn't work for test configs that use a prereq config. Yes, if the test config fails to apply, the except clause runs the destroy, but then it exits without cleaning up the prereq config. Maybe I'm wrong, but this is the logic branch I'm seeing:

1. Apply prereq config
2. Apply test config - fail!
3. Except clause destroys test config
4. Except clause exits
5. Prereq config remains behind

Reply:
You're correct, the prereq won't get destroyed if the test config fails.  I'll look into a change.

Reply:
ok, I've made a change.  Normally pytest is easy to work with, but when you need a function that is both a fixture and requires a "dynamic" argument (versus a test parameter), it's ugly.

Reply:
I can follow the new code and it _looks_ alright as far as ordering the workflow, but I don't love it. I'll approve it, but maybe as we settle into it more, an approach will become more clear that allows us to make the fixtures more static and use standard pytest teardown features here.

Reply:
The problem is that we're running two actions, each with their own teardown and with one dependent on another, as one test.  The dependency is the issue.  It's not the normal way of testing.  Plus the fact that we have fixtures that take dynamic parameters.

I can look at it some more, but I feel swamped with all the side issues going on (unresolved test failures, changes to file_cache.py, changes to this code, unresolved changes to the new Travis work flow).  All of the potential problems with this automated test have not been discovered, which is I was trying to focus on, but I keep getting distracted by the side issues.

This work should be broken down into smaller, more manageable chunks.  To that affect, I will create some Trello tasks.

Reply:
Yep, I get it. I was hopeful that the `yield` option would work, as the docs describe it tracks this ordering requirement just the way we need, more or less as a FILO queue. Maybe if we can figure out a more declarative way to setup the test framework it will be less dynamic, and the standard pytest features will just work. But that's a future task for sure. I am trying to help you feel _less_ swamped by this set of interrelated changes, not more!